### ISSUE TITLE
 invalid memory address or nil pointer dereference at (*iRunner).getDeliverFunc

### ISSUE THREAD

**Issue by sathieu** [mozilla-services/heka#1825]
(Using 0.10 at 90eea62b35699, i.e. almost 0.10 release).

Log excerp
```
[Heka starting 183 apache access + error inputs, and some more]
2015/12/29 11:35:32 Input started: <...>
2015/12/29 11:42:25 Diagnostics: 32 packs have been idle more than 120 seconds.
2015/12/29 11:42:25 Diagnostics: (input) Plugin names and quantities found on idle packs:
2015/12/29 11:42:25 Diagnostics:        TcpOutput: 32
2015/12/29 11:42:25
[... more idle packs ...]
2015/12/29 15:06:55 Diagnostics: 32 packs have been idle more than 120 seconds.
2015/12/29 15:06:55 Diagnostics: (input) Plugin names and quantities found on idle packs:
2015/12/29 15:06:55 Diagnostics:        TcpOutput: 32
2015/12/29 15:06:55
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x70 pc=0x4a5efe]

goroutine 34 [running]:
github.com/mozilla-services/heka/pipeline.(*iRunner).getDeliverFunc(0xc208873040, 0xc20a29b650, 0x1, 0xc200000000, 0x0, 0x0, 0x0, 0x0)
        /root/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:453 +0x5be
github.com/mozilla-services/heka/pipeline.(*iRunner).NewDeliverer(0xc208873040, 0xc20a29b650, 0x1, 0x0, 0x0)
        /root/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:514 +0x5a
github.com/mozilla-services/heka/plugins/logstreamer.(*LogstreamerInput).startLogstreamInput(0xc20839b810, 0xc20a072600, 0x1, 0x7fa29dc94d58, 0xc208873040, 0x7fa29dc94970, 0xc208075000)
        /root/heka/build/heka/src/github.com/mozilla-services/heka/plugins/logstreamer/logstreamer_input.go:184 +0x8e
github.com/mozilla-services/heka/plugins/logstreamer.(*LogstreamerInput).Run(0xc20839b810, 0x7fa29dc94d58, 0xc208873040, 0x7fa29dc94970, 0xc208075000, 0x0, 0x0)
        /root/heka/build/heka/src/github.com/mozilla-services/heka/plugins/logstreamer/logstreamer_input.go:246 +0x9ab
github.com/mozilla-services/heka/pipeline.(*iRunner).Starter(0xc208873040, 0x7fa29dc94970, 0xc208075000, 0xc2080751a0)
        /root/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:325 +0x19e
created by github.com/mozilla-services/heka/pipeline.(*iRunner).Start
        /root/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:305 +0x7e7

[492 more goroutines]
```
Putting this for reference.

**Comment by sathieu** [comment]
Applied https://gist.github.com/sathieu/6e30382759f0ee0e1b65032c26ab7d10#file-heka1825-diff

Now I'm getting (at start):

```
sudo -u heka hekad -config /etc/heka/conf.d/
2016/05/11 16:39:30 Pre-loading: [ModSecuritySplitter]
2016/05/11 16:39:30 Pre-loading: [ModSecurityDecoder]
2016/05/11 16:39:30 Pre-loading: [ApacheAccessNantesInput]
2016/05/11 16:39:30 Pre-loading: [ApacheAccessNantesDecoder]
2016/05/11 16:39:30 Pre-loading: [ApacheAccessNantesSubDecoder]
2016/05/11 16:39:30 Pre-loading: [ApacheErrorInput]
2016/05/11 16:39:30 Pre-loading: [ApacheErrorDecoder]
2016/05/11 16:39:30 Pre-loading: [ModSecurityInput]
2016/05/11 16:39:30 Pre-loading: [auth_log]
2016/05/11 16:39:30 Pre-loading: [LinuxAuthlogDecoder]
2016/05/11 16:39:30 Pre-loading: [syslog]
2016/05/11 16:39:30 Pre-loading: [LinuxSyslogDecoder]
2016/05/11 16:39:30 Pre-loading: [GeoipRemote_addrDecoder]
2016/05/11 16:39:30 Pre-loading: [TcpOutput]
2016/05/11 16:39:30 Pre-loading: [ProtobufDecoder]
2016/05/11 16:39:30 Loading: [ProtobufDecoder]
2016/05/11 16:39:30 Pre-loading: [ProtobufEncoder]
2016/05/11 16:39:30 Loading: [ProtobufEncoder]
2016/05/11 16:39:30 Pre-loading: [TokenSplitter]
2016/05/11 16:39:30 Loading: [TokenSplitter]
2016/05/11 16:39:30 Pre-loading: [HekaFramingSplitter]
2016/05/11 16:39:30 Loading: [HekaFramingSplitter]
2016/05/11 16:39:30 Pre-loading: [NullSplitter]
2016/05/11 16:39:30 Loading: [NullSplitter]
2016/05/11 16:39:30 Loading: [ModSecurityDecoder]
2016/05/11 16:39:30 Loading: [ApacheAccessNantesSubDecoder]
2016/05/11 16:39:30 Loading: [ApacheErrorDecoder]
2016/05/11 16:39:30 Loading: [LinuxAuthlogDecoder]
2016/05/11 16:39:30 Loading: [LinuxSyslogDecoder]
2016/05/11 16:39:30 Loading: [GeoipRemote_addrDecoder]
2016/05/11 16:39:30 Loading: [ApacheAccessNantesDecoder]
2016/05/11 16:39:30 Loading: [ModSecuritySplitter]
2016/05/11 16:39:30 Loading: [ApacheAccessNantesInput]
2016/05/11 16:39:30 Loading: [ApacheErrorInput]
2016/05/11 16:39:30 Loading: [ModSecurityInput]
2016/05/11 16:39:30 Loading: [auth_log]
2016/05/11 16:39:30 Loading: [syslog]
2016/05/11 16:39:31 Loading: [TcpOutput]
2016/05/11 16:39:31 Starting hekad...
2016/05/11 16:39:31 Output started: TcpOutput
2016/05/11 16:39:31 MessageRouter started.
2016/05/11 16:39:31 Input started: ApacheErrorInput
2016/05/11 16:39:31 Input started: ModSecurityInput
2016/05/11 16:39:31 Input started: auth_log
2016/05/11 16:39:31 Input started: syslog
2016/05/11 16:39:31 Input started: ApacheAccessNantesInput
2016/05/11 16:39:32 Input 'ApacheAccessNantesInput' error: Unable to launch decoder 'ApacheAccessNantesDecoder'
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4a3b7c]

goroutine 204 [running]:
github.com/mozilla-services/heka/pipeline.(*deliverer).Deliver(0xc209a0fb60, 0xc2081f7900)
    /root/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:131 +0x2c
github.com/mozilla-services/heka/pipeline.(*sRunner).DeliverRecord(0xc209a414a0, 0xc209a42000, 0x14e, 0x2000, 0x7fabde8604f0, 0xc209a0fb60)
    /root/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/splitter_runner.go:286 +0x367
github.com/mozilla-services/heka/plugins/logstreamer.(*LogstreamInput).deliverRecords(0xc208124a00, 0x0, 0x0)
    /root/heka/build/heka/src/github.com/mozilla-services/heka/plugins/logstreamer/logstreamer_input.go:378 +0x486
github.com/mozilla-services/heka/plugins/logstreamer.(*LogstreamInput).Run(0xc208124a00, 0x7fabde857208, 0xc208044000, 0x7fabde856888, 0xc208076a00, 0xc209b84060, 0x7fabde8604f0, 0xc209a0fb60, 0x7fabde860528, 0xc209a414a0)
    /root/heka/build/heka/src/github.com/mozilla-services/heka/plugins/logstreamer/logstreamer_input.go:316 +0x204
created by github.com/mozilla-services/heka/plugins/logstreamer.(*LogstreamerInput).startLogstreamInput
    /root/heka/build/heka/src/github.com/mozilla-services/heka/plugins/logstreamer/logstreamer_input.go:196 +0x291

[...]
```

**Comment by sathieu** [comment]
Here is the relevant part of my config:

```
[ApacheAccessNantesInput]
type = "LogstreamerInput"
log_directory = '/var/log/apache2'
file_match = '(?P<servername>[^/]+)/access-(?P<environment_appli>[^.-]+)(-(?P<Year>\d{4})(?P<Month>\d{2})(?P<Day>\d{2}))?\.log'
priority = ["Year", "Month", "Day"]
decoder = "ApacheAccessNantesDecoder"
differentiator = ["apache_access-", "environment_appli", "-", "servername"]
send_decode_failures = true
log_decode_failures = false
[ApacheAccessNantesInput.translation.Year]
missing = 9999

[ApacheAccessNantesDecoder]
type = "MultiDecoder"
subs = ['ApacheAccessNantesSubDecoder', 'GeoipRemote_addrDecoder']
cascade_strategy = "all"

[ApacheAccessNantesSubDecoder]
type = "SandboxDecoder"
filename = "lua_decoders/apache_access.lua"
[ApacheAccessNantesSubDecoder.config]
type = "apache_access_nantes"
user_agent_transform = true
user_agent_keep = true
log_format = '%v %h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\" %I %O \"%{Location}i\" \"%{X-Forwarded-For}i\" %D %p \"%{SSL_PROTOCOL}x\" \"%{SSL_CIPHER}x\"'
```

**Comment by sathieu** [comment]
Setting this helps:

```
subs = ['ApacheAccessNantesSubDecoder'] # i.e removing GeoIP
```

So this may be related to geoip (or some threshold hit only when geoip is loaded)

**Comment by sathieu** [comment]
I'll submit a patch to improve error reporting, with it I have:

```
2016/05/12 11:36:50 Input 'ApacheAccessNantesInput' error: Initialization failed for 'ApacheAccessNantesDecoder': Non-existent subdecoder: GeoipRemote_addrDecoder
(then panic)
```

**Comment by sathieu** [comment]
One more step:

```
Input 'ApacheAccessNantesInput' error: Initialization failed for 'ApacheAccessNantesDecoder': Non-existent subdecoder: GeoipRemote_addrDecoder (Initialization failed for 'GeoipRemote_addrDecoder': Could not open GeoIP database: %!s(MISSING)
```

(oups!)

**Comment by sathieu** [comment]
Now:

```
Input 'ApacheAccessNantesInput' error: Initialization failed for 'ApacheAccessNantesDecoder': Non-existent subdecoder: GeoipRemote_addrDecoder (Initialization failed for 'GeoipRemote_addrDecoder': Could not open GeoIP database '/usr/local/share/geoip/GeoLiteCity.dat': Error opening GeoIP database ([/usr/local/share/geoip/GeoLiteCity.dat]): cannot allocate memory)
```

**Comment by sathieu** [comment]
it looks like the geoip decoder is instantiated too many times.

**Comment by sathieu** [comment]
I've implemented GeoIP flags: #1935. I hope it'll help.

**Comment by sathieu** [comment]
Summary:

I've implemented:
- Improve error reporting when Decoder initialization failed (dev=#1934, 0.10=#1932)
- Support GeoIP flags (dev=#1935, 0.10=TODO)

I'm now using flags="GEOIP_MMAP_CACHE", and it works.

So, I'm closing this bug. Please note that the panic is still here, but just after the error message which points to the solution.
